### PR TITLE
updated play to conform to diagram

### LINKED ISSUE #920: Clean up play
[See diagram](https://app.diagrams.net/#G1DpxnUs13rEW11IShEQgp5BxpmRpTWRBH)

### PR DESCRIPTION
**Pull request by RobotJente**
The only difference between the diagram (besides the TODO's you see in the code, which are waiting for the role refactor) is that I have chosen to check whether the play is finished in ApplicationManager, because this seemed more logical to me than having an invariant to check if all the roles are not finished and having to call that in every derived class

### CODE REVIEW

File: src/stp/new_plays/TestPlay.cpp [hunk 1]
```diff
@@ -62,7 +63,15 @@ void TestPlay::calculateInfoForRoles() noexcept {
                                                        world->getWorld()->getBall()->get()->getFilteredVelocity() * 0.5);
             }
         }
+
```
Comment (RobotJente) [inline src/stp/new_plays/TestPlay.cpp]:
```suggestion
```

File: src/stp/new_roles/TestRole.cpp [hunk 2]
```diff
@@ -7,13 +7,14 @@
 #include <stp/new_tactics/TestTactic.h>
 
 #include <utility>
+#include <include/roboteam_ai/stp/new_tactics/GetBall.h>
 
 namespace rtt::ai::stp {
 
 TestRole::TestRole(std::string name) : Role(std::move(name)) {
+
```
Comment (RobotJente) [inline src/stp/new_roles/TestRole.cpp]:
```suggestion
```

File: include/roboteam_ai/stp/Play.hpp [hunk 3]
```diff
@@ -41,7 +41,7 @@ class Play {
      * if any waiting -> waiting
      * otherwise -> running
      */
-    [[nodiscard]] virtual Status update() noexcept;
+    [[nodiscard]] virtual void update() noexcept;
```
Comment (timovdk) [inline include/roboteam_ai/stp/Play.hpp]: void doesn't need nodiscard
```suggestion
    virtual void update() noexcept;
```

File: include/roboteam_ai/stp/Play.hpp [hunk 4]
```diff
@@ -103,11 +123,12 @@ class Play {
      */
     rtt::ai::Field field;
 
-    protected:
     /**
      * Assigns robots to roles
      */
     virtual void assignRoles() noexcept = 0;
+
+    bool shouldRoleSkipEndTactic();
```
Comment (timovdk) [inline include/roboteam_ai/stp/Play.hpp]: This function should probably have a comment

File: include/roboteam_ai/stp/new_plays/Pass.h [hunk 5]
```diff
@@ -16,6 +16,10 @@ class Pass : public Play {
      */
     Pass();
 
+    bool isValidPlayToStart(world_new::World *world) override;
+
+    bool isValidPlayToKeep(world_new::World *world) override;
```
Comment (timovdk) [inline include/roboteam_ai/stp/new_plays/Pass.h]: comments

File: src/stp/Play.cpp [hunk 6]
```diff
@@ -16,51 +16,46 @@ void Play::updateWorld(world_new::World* world) noexcept {
...
-    }
+    // derived class method call
+    calculateInfoForRoles();
 
-    if (count[static_cast<size_t>(Status::Failure)]) {
-        return Status::Failure;
+    for (auto& role : roles) {
+        // Update the roles
+        auto roleStatus = role->update(stpInfos[role->getName()]);
+        roleStatuses.emplace_back(roleStatus);
```
Comment (timovdk) [inline src/stp/Play.cpp:54]: Won't roleStatuses get infinitely big?

File: src/stp/Play.cpp [hunk 7]
```diff
@@ -16,51 +16,46 @@ void Play::updateWorld(world_new::World* world) noexcept {
...
+            }
+        }
     }
+}
 
-    if (count[static_cast<size_t>(Status::Waiting)]) {
-        return Status::Waiting;
+    bool Play::shouldRoleSkipEndTactic() {
+        return false;
     }
```
Comment (timovdk) [inline src/stp/Play.cpp]: Should this function be implemented  here? or should it be purely virtual

File: src/stp/new_plays/TestPlay.cpp [hunk 8]
```diff
@@ -52,6 +52,7 @@ void TestPlay::assignRoles() noexcept {
     }
 }
 
+
```
Comment (timovdk) [inline src/stp/new_plays/TestPlay.cpp]:
```suggestion
```

File: src/stp/new_roles/TestRole.cpp [hunk 9]
```diff
@@ -7,13 +7,13 @@
 #include <stp/new_tactics/TestTactic.h>
 
 #include <utility>
+#include <include/roboteam_ai/stp/new_tactics/GetBall.h>
```
Comment (timovdk) [inline src/stp/new_roles/TestRole.cpp]:
```suggestion
#include <stp/new_tactics/GetBall.h>
```

File: src/stp/Play.cpp [hunk 10]
```diff
@@ -16,51 +16,55 @@ void Play::updateWorld(world_new::World* world) noexcept {
     this->field = world->getField().value();
 }
 
-Status Play::update() noexcept {
-    constexpr static size_t ENUM_COUNT = 4;
-    std::array<size_t, ENUM_COUNT> count{};
-    std::fill(count.begin(), count.end(), 0);
+void Play::update() noexcept {
+    roleStatuses.clear();
```
Comment (timovdk) [inline src/stp/Play.cpp:21]: maybe a comment as it is not clear why this is done

File: include/roboteam_ai/stp/Play.hpp [hunk 11]
```diff
@@ -81,12 +69,35 @@ class Play {
      */
     Play(Play&& other) = default;
 
+    /**
+     * Check if the preconditions of this play are true
+     * @return true if the play is allowed to be started, else false
+     */
+    [[nodiscard]] virtual bool isValidPlayToStart(world_new::World* world) noexcept = 0;
```
Comment (Ratoone) [inline include/roboteam_ai/stp/Play.hpp:76]: Reference instead of pointer

File: test/StpTests/PlayCheckerTests.cpp [hunk 12]
```diff
@@ -20,13 +16,14 @@ class AlwaysValid : public rtt::ai::stp::Play {
     }
     virtual void calculateInfoForRoles() noexcept {
     }
+
+    bool isValidPlayToStart(rtt::world_new::World *world) noexcept override{return true;}
+    bool isValidPlayToKeep(rtt::world_new::World *world) noexcept override{return true;}
+    bool shouldRoleSkipEndTactic() override { return false;}
+
```
Comment (Ratoone) [inline test/StpTests/PlayCheckerTests.cpp]: This whole thing can be done cleaner by mocking these methods. Then you don't have to implement a new class for each situation

File: include/roboteam_ai/stp/Play.hpp [hunk 13]
```diff
@@ -103,11 +114,17 @@ class Play {
...
      * Assigns robots to roles
      */
     virtual void assignRoles() noexcept = 0;
+
+    /**
+     * This function is used to determine if, when a role is in an endtactic, that endtactic should be skipped.
+     * An example could be BlockRobot and Intercept. You block a robot until a ball is shot and then the robot
+     * closest to the ball should try to intercept
+     */
+    virtual bool shouldRoleSkipEndTactic() = 0;
```
Comment (Ratoone) [inline include/roboteam_ai/stp/Play.hpp:127]: I feel this needs more information. The world? The current robot? The role that you apply this to? Something

**Review by Ratoone — CHANGES_REQUESTED** [review]
Learn to mock boi

File: include/roboteam_ai/stp/Play.hpp (hunk 13 above)
Comment (RobotJente) [reply include/roboteam_ai/stp/Play.hpp:127]: Probably, but if it does you can just add it right?

File: src/stp/Play.cpp (hunk 10 above)
Comment (RobotJente) [reply src/stp/Play.cpp:21]:
```suggestion
    // clear roleStatuses so it only contains the current tick's statuses
    roleStatuses.clear();
```

File: src/stp/new_plays/Pass.cpp [hunk 14]
```diff
@@ -35,14 +34,14 @@ void Pass::assignRoles() noexcept {
...
-    flagMap.insert({"test_role_9", {closeToBallFlag}});
-    flagMap.insert({"test_role_10", {closeToTheirGoalFlag}});
+//    flagMap.insert({"test_role_3", {closeToTheirGoalFlag}});
+//    flagMap.insert({"test_role_4", {closeToBallFlag}});
+//    flagMap.insert({"test_role_5", {closeToTheirGoalFlag, closeToBallFlag}}
+//    flagMap.insert({"test_role_6", {closeToBallFlag}});
+//    flagMap.insert({"test_role_7", {closeToTheirGoalFlag}});
+//    flagMap.insert({"test_role_8", {closeToTheirGoalFlag, closeToBallFlag}});
+//    flagMap.insert({"test_role_9", {closeToBallFlag}});
+//    flagMap.insert({"test_role_10", {closeToTheirGoalFlag}});
```
Comment (RobotJente) [inline src/stp/new_plays/Pass.cpp]:
```suggestion
    flagMap.insert({"test_role_3", {closeToTheirGoalFlag}});
    flagMap.insert({"test_role_4", {closeToBallFlag}});
    flagMap.insert({"test_role_5", {closeToTheirGoalFlag, closeToBallFlag}}
    flagMap.insert({"test_role_6", {closeToBallFlag}});
    flagMap.insert({"test_role_7", {closeToTheirGoalFlag}});
    flagMap.insert({"test_role_8", {closeToTheirGoalFlag, closeToBallFlag}});
    flagMap.insert({"test_role_9", {closeToBallFlag}});
    flagMap.insert({"test_role_10", {closeToTheirGoalFlag}});

```

File: test/StpTests/PlayCheckerTests.cpp (hunk 12 above)
Comment (RobotJente) [reply test/StpTests/PlayCheckerTests.cpp]: Probably. I didn't write these tests, I'll make it an issue though

File: test/StpTests/PlayCheckerTests.cpp [hunk 15]
```diff
@@ -36,10 +33,16 @@ class AlwaysFalse : public rtt::ai::stp::Play {
 
     virtual void calculateInfoForRoles() noexcept {
     }
+
+    bool isValidPlayToStart(rtt::world_new::World *world) noexcept override{return false;}
+    bool isValidPlayToKeep(rtt::world_new::World *world) noexcept override{return false;}
+    bool shouldRoleSkipEndTactic() override { return false;}
+
```
Comment (RobotJente) [inline test/StpTests/PlayCheckerTests.cpp:27]:
```suggestion
```

File: test/StpTests/PlayCheckerTests.cpp [hunk 16]
```diff
@@ -36,10 +33,16 @@ class AlwaysFalse : public rtt::ai::stp::Play {
...
 };
 
 class AnotherAlwaysTrue : public AlwaysValid{
     using AlwaysValid::AlwaysValid;
+
```
Comment (RobotJente) [inline test/StpTests/PlayCheckerTests.cpp]:
```suggestion
```

File: test/StpTests/PlayDeciderTests.cpp [hunk 17]
```diff
@@ -9,9 +9,6 @@
 
 class AlwaysValid : public rtt::ai::stp::Play {
 public:
-    bool isValidPlay(rtt::world_new::World *world) noexcept override {
-        return true;
-    }
```
Comment (RobotJente) [inline test/StpTests/PlayDeciderTests.cpp:13]:
```suggestion
```

File: test/StpTests/PlayDeciderTests.cpp [hunk 18]
```diff
@@ -22,12 +19,13 @@ class AlwaysValid : public rtt::ai::stp::Play {
 
     virtual void calculateInfoForRoles() noexcept {
     }
+
+    bool isValidPlayToStart(rtt::world_new::World *world) noexcept override{return true;}
+    bool isValidPlayToKeep(rtt::world_new::World *world) noexcept override{return true;}
+    bool shouldRoleSkipEndTactic() override { return false;}
 };
```
Comment (RobotJente) [inline test/StpTests/PlayDeciderTests.cpp:22]:
```suggestion
```

File: include/roboteam_ai/stp/Play.hpp (hunk 11 above)
Comment (RobotJente) [reply include/roboteam_ai/stp/Play.hpp:76]: World is managed so it shouldn't really matter, right?

File: test/StpTests/PlayCheckerTests.cpp [hunk 19]
```diff
@@ -36,6 +33,10 @@ class AlwaysFalse : public rtt::ai::stp::Play {
 
     virtual void calculateInfoForRoles() noexcept {
     }
+
+    bool isValidPlayToStart(rtt::world_new::World *world) noexcept override{return false;}
+    bool isValidPlayToKeep(rtt::world_new::World *world) noexcept override{return false;}
+    bool shouldRoleSkipEndTactic() override { return false;}
 };
```
Comment (RobotJente) [inline test/StpTests/PlayCheckerTests.cpp:32]:
```suggestion

```

Comment (RobotJente) [inline test/StpTests/PlayCheckerTests.cpp:32]:
```suggestion
```

File: src/stp/new_plays/Pass.cpp [hunk 20]
```diff
@@ -37,7 +36,7 @@ void Pass::assignRoles() noexcept {
     flagMap.insert({"defender1", {notImportant}});
     flagMap.insert({"test_role_3", {closeToTheirGoalFlag}});
     flagMap.insert({"test_role_4", {closeToBallFlag}});
-    flagMap.insert({"test_role_5", {closeToTheirGoalFlag, closeToBallFlag}});
+    flagMap.insert({"test_role_5", {closeToTheirGoalFlag, closeToBallFlag}}
```
Comment (timovdk) [inline src/stp/new_plays/Pass.cpp]:
```suggestion
    flagMap.insert({"test_role_5", {closeToTheirGoalFlag, closeToBallFlag}});
```

File: src/stp/new_plays/Pass.cpp [hunk 21]
```diff
@@ -37,7 +36,7 @@ void Pass::assignRoles() noexcept {
     flagMap.insert({"defender1", {notImportant}});
     flagMap.insert({"test_role_3", {closeToTheirGoalFlag}});
     flagMap.insert({"test_role_4", {closeToBallFlag}});
-    flagMap.insert({"test_role_5", {closeToTheirGoalFlag, closeToBallFlag}});
+    flagMap.insert({"test_role_5", {closeToTheirGoalFlag, closeToBallFlag}};
```
Comment (timovdk) [inline src/stp/new_plays/Pass.cpp]:
```suggestion
    flagMap.insert({"test_role_5", {closeToTheirGoalFlag, closeToBallFlag}});
```

**Review by timovdk — APPROVED** [review]

**Pull request merged**